immuni_analytics/apis/analytics.py

```diff
@@ -0,0 +1,276 @@
+    Base64String,
+    IntegerBoolField,
+    IsoDate,
+    Province,
+)
+from immuni_common.models.swagger import HeaderImmuniContentTypeJson
+
+_LOGGER = logging.getLogger(__name__)
+
+bp = Blueprint("analytics", url_prefix="analytics")
+
+
+@bp.route("/apple/operational-info", methods=["POST"], version=1)
+@doc.summary("Upload operational info (caller: Mobile Client.)")
+@doc.description(
+    "Check if the analytics_token is authorized to upload and save the operational info"
+    " in the database."
+)
+@doc.consumes(doc.Boolean(name="Immuni-Dummy-Data", required=True), location="headers")
+@doc.consumes(AppleOperationalInfo, location="body")
+@doc.consumes(HeaderImmuniContentTypeJson(), location="header", required=True)
+@doc.consumes(
+    doc.String(name="Authorization", description="Bearer <ANALYTICS_TOKEN>"),
+    location="header",
+    required=True,
+)
+@doc_exception(SchemaValidationException)
+@doc.response(
+    HTTPStatus.NO_CONTENT.value, None, description="Well-formed request.",
+)
+@validate(
+    location=Location.HEADERS,
+    is_dummy=IntegerBoolField(
+        required=True,
+        data_key="Immuni-Dummy-Data",
+        allow_strings=True,
+        description="Whether the current request is a dummy request. Dummy requests are ignored.",
+    ),
+)
+@validate(
+    location=Location.JSON,
+    province=Province(),
+    exposure_permission=IntegerBoolField(required=True),
+    bluetooth_active=IntegerBoolField(required=True),
+    notification_permission=IntegerBoolField(required=True),
+    exposure_notification=IntegerBoolField(required=True),
+    last_risky_exposure_on=IsoDate(),
+)
+async def post_operational_info(  # pylint: disable=too-many-arguments
+    request: Request,
+    is_dummy: bool,
+    province: str,
+    exposure_permission: bool,
+    bluetooth_active: bool,
+    notification_permission: bool,
+    exposure_notification: bool,
+    last_risky_exposure_on: date,
+) -> HTTPResponse:
+    """
+    Check if the analytics_token is authorized and save the operational data
+    in compliance with the rate limiting policy.
+
+    :param request: the HTTP request object.
+    :param is_dummy: whether the uploaded data is dummy or not.
+    :param province: the user's province of residence.
+    :param exposure_permission: whether the user has granted the App permission to use the
+     Exposure Notifications framework.
+    :param bluetooth_active: whether the user has Bluetooth turned on.
+    :param notification_permission: whether the user has granted the App permission to display
+     notifications.
+    :param exposure_notification: whether the user was notified of a Risky Exposure after the
+     last Exposure Detection.
+    :param last_risky_exposure_on: the date on which the last Risky Exposure took place, if any.
+    :return: 204 in any case.
+    """
+    if is_dummy:
+        return json_response(body=None, status=HTTPStatus.NO_CONTENT)
+
+    if await managers.analytics_redis.srem(
```

Comment:
🤔  If the client sends the request too early (before the token is processed by celery) then nobody notices. We might want to send a different response if the token is invalid (like 401 ?)

Also, it might be worth to add two metrics: One for all of the unauthorized requests and one for all of the unused tokens (although I'm not sure how you could count them), and see if they match (and make sure they're low).

Reply:
The client must call the token endpoint first to check if their token is valid.

Reply:
I would add all the metrics together when the code is stable.